components/AuthProvider.tsx

```diff
@@ -0,0 +1,75 @@
+import { UserDoc } from 'models/users'
+import { onAuthStateChanged, User } from 'firebase/auth'
+import { doc, onSnapshot } from 'firebase/firestore'
+import React, {
+  createContext,
+  ReactNode,
+  useContext,
+  useEffect,
+  useState,
+} from 'react'
+import { auth, userRef } from 'utils/firebase'
+
+const authContext = createContext<{
+  user: User | null
+  userLoading: boolean
+  userData: UserDoc | null
+  userDataLoading: boolean
+}>({
+  user: null,
+  userLoading: true,
+  userData: null,
+  userDataLoading: true,
+})
+
+export const useUser = () => {
+  return useContext(authContext)
+}
+
+//There is a distinction between user and userData
+//User is the data used by Firebase to distinguish different user accounts
+//userData is the data we use to handle list ownership
+
+const AuthProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
+  const [user, setUser] = useState<User | null>(null)
+  const [userLoading, setLoading] = useState(true)
+  const [userData, setUserData] = useState<UserDoc | null>(null)
+  const [userDataLoading, setUserDataLoading] = useState(true)
+
+  useEffect(() => {
+    setUserDataLoading(true)
+    if (user) {
+      const unsub = onSnapshot(doc(userRef, user.uid), (doc) => {
+        const data = doc.data() as UserDoc
```

Comment:
Can we type something somewhere where we don't have to cast this?

Reply:
I just looked into it and there is a way to convert `doc.data()` to the correct type without casting but I'm honestly having a hard time trying to implement it. I'll look further into it and see what I can do. 

Heres a link if you're interested:
https://firebase.google.com/docs/reference/node/firebase.firestore.FirestoreDataConverter